### ISSUE TITLE
Not working with native gradle locks

### ISSUE THREAD

**Issue by slavashvets** [spring-gradle-plugins/dependency-management-plugin#258]
**Gradle version:** 5.4.1
**Plugin version:** 1.0.8.RELEASE

## Steps to reproduce

Create build.gradle
```gradle
configurations {
  check
}

dependencies {
  check group: 'org.dummy', name: 'dummy'
}

repositories {
  mavenCentral()
}

task check {
  doLast {
    configurations.findAll {
      it.canBeResolved
    }.each {
      it.resolve().each {
        println(it.name)
      }
    }
  }
}

dependencyLocking {
  lockAllConfigurations()
}
task lock {
  doFirst {
    assert gradle.startParameter.writeDependencyLocks
  }
  doLast {
    configurations.findAll {
      it.canBeResolved
    }.each { it.resolve() }
  }
}

buildscript {
  repositories {
    maven {
      url "https://plugins.gradle.org/m2/"
    }
  }
  dependencies {
    classpath "io.spring.gradle:dependency-management-plugin:1.0.8.RELEASE"
  }
}

apply plugin: "io.spring.dependency-management"

dependencyManagement {
  dependencies {
    dependency 'org.dummy:dummy:1.+'
  }
}
```

Generate lockfile (assume the locked version is *1.1*):
`./gradlew lock --write-locks`

Upload new version of "org.dummy:dummy" to remote repository (e.g. *1.2*)

Run build
`./gradlew check`

## Expected
No errors

## Actual
```
$ ./gradlew check
> Task :check FAILED

FAILURE: Build failed with an exception.

* Where:
Build file 'build.gradle' line: 19

* What went wrong:
Execution failed for task ':check'.
> Could not resolve all files for configuration ':compile'.
   > Could not resolve org.dummy:dummy:{strictly 1.1}.
     Required by:
         project :
      > Cannot find a version of 'org.dummy:dummy' that satisfies the version constraints:
           Constraint path ':reproduce-gradle-lock-bug:unspecified' --> 'org.dummy:dummy:{strictly 1.1}' because of the following reason: dependency was locked to version '1.1'
```

**Comment by wilkinsona** [comment]
Interesting. Thanks for the detailed report. Can you expand a bit on the expected behaviour please? What version of `org.dummy:dummy` do you expect to be used? I assume 1.1, but I'd like to be certain.

**Comment by slavashvets** [comment]
@wilkinsona Yes, I expect to see version 1.1 downloaded because it's in the lockfile. However, I see the error like if I write exact version ```dependency 'org.dummy:dummy:1.2```

**Comment by wilkinsona** [comment]
I've been unable to reproduce the failure, but I have reproduced something that isn't behaving as expected. With `org.springframework:spring-jcl` locked to 5.1.8.RELEASE, 5.1.8 is used when the dependency is declared like this:

```
dependencies {
  check group: 'org.springframework', name: 'spring-jcl', version: '5.1.+'
}
```

However, 5.1.9 is used when the version is specified using dependency management:

```
dependencies {
  check group: 'org.springframework', name: 'spring-jcl'
}

dependencyManagement {
  dependencies {
    dependency 'org.springframework:spring-jcl:5.1.+'
  }
}
```

So, it appears that the locking is being overridden rather than build failing due to an unsatisfiable version constraint.

**Comment by NielsDoucet** [comment]
We've hit this issue as well.
Is there a workaround available, or is a fix planned in the near future?

**Comment by wilkinsona** [comment]
I hope to fix the issue, but I don't yet know how to do so. It requires identifying when a dependency version has been locked and, therefore, should be ignored by dependency management. I don't know how to do that yet, or if it's even possible.

You may want to investigate using Gradle's build-in support for handling Maven mom's and dependency constraints as I suspect it won't suffer from the same problem.

**Comment by NielsDoucet** [comment]
We tried that, but there are some cases where this leads to issues due to the differences between maven's dependencyManagement model and gradle's platform (as you outlined very nicely).
So for some of our simpler projects this works, but it's not a global solution for us.

**Comment by wilkinsona** [comment]
There doesn't appear to be a specific Gradle API for identifying a locked dependency, however it can be inferred reasonably accurately by looking for a strict version constraint. However, this appears to only be possible by resorting to reflection:

```java
        try {
            Field delegateField = details.getClass().getDeclaredField("delegate");
            delegateField.setAccessible(true);
            DependencySubstitution dependencySubstitution = (DependencySubstitution) delegateField.get(details);
            ComponentSelector selector = dependencySubstitution.getRequested();
            Class<?> selectorClass = selector.getClass();
            Method method = selectorClass.getMethod("getVersionConstraint");
            Object versionConstraint = method.invoke(selector);
            String strictVersion = (String)versionConstraint.getClass().getMethod("getStrictVersion").invoke(versionConstraint);
            if (strictVersion != null) {
                logger.debug("'{}' has a strict version constraint. Dependency management has not been applied", details.getRequested());
                return;
            }
        }
        catch (Exception ex) {
            // Best-effort attempt at honoring strict version constraint failed. Continue.
        }
```

In the above, `details` is an instance of `DependencyResolveDetails`. If this plugin's minimum version of Gradle was raised it would be possible to avoid the reflection around `VersionConstraint`, but I think some would still be necessary to bridge across from the `DependencyResolveDetails` to `DependencySubsitution`.

@melix once the Gradle 6.0 push is over I'd be most grateful for you help here. Specifically, I'm wondering if there is a way, given a `DependencyResolveDetails` to identify that the dependency has been locked using public API.

**Comment by melix** [comment]
You can't really infer that _one dependency_ was locked, but you can certainly know that a _configuration_ was locked. If a configuration is locked, then by definition all its dependencies are (well, except when in a dependency lock update mode).

Unfortunately, unless I'm mistaken, there's no public API to do this (@ljacomet am I correct?). So you can just do this:

```
import org.gradle.api.internal.artifacts.configurations.ResolutionStrategyInternal;

ResolutionStrategyInternal resolutionStrategy = (ResolutionStrategyInternal) myConf.getResolutionStrategy();

if (resolutionStrategy.isDependencyLockingEnabled()) {
   ...
}
```

**Comment by ljacomet** [comment]
There is indeed no public API at this point. And the _locked_ status might not be enough, because a locked configuration may not have any lock state at all.

I think we may want to use this plugin's use case to think about what kind of API dependency management related plugins would need.

**Comment by wilkinsona** [comment]
Thank you, @melix and @ljacomet.

For the use case that's described here, I think what's needed is for it to be possible to detect if a lock is in place and to then honour it by excluding the dependency from version management. I guess things may get a bit more complicated if the dependency management isn't compatible with the lock. In the example above, the dependency management says 5.1.+. Perhaps a failure should result if the dependency is locked to a 5.0.x version.

**Comment by NfNitLoop** [comment]
I've pulled the Spring Dependency Management plugin out of a project because of this issue.  (I think?)

A new version of ElasticSearch was being installed. That version was too new to be compatible with the server I'm talking to.
So I followed Gradle documentation to set a dependency constraint:

```kotlin
dependencies {
    // [my dependencies]
    
    constraints {
        implementation("org.elasticsearch.client:elasticsearch-rest-high-level-client") {
            version {
                strictly("[7.10.2,7.11)")
            }
        }
    }
}
```

The SDM plugin was completely ignoring my constraints. And looking at the output of `gradle dependencyInsight` for the dependency didn't really make it clear *why* it was choosing a newer version than any of my direct/transitive dependencies.

I was only using the SDM plugin because Spring Initializr generated a project skeleton with it. But if it results in a gradle config that ignores my explicit versions, maybe they shouldn't?

**Comment by wilkinsona** [comment]
Sorry about that, @NfNitLoop. It’s https://github.com/spring-gradle-plugins/dependency-management-plugin/issues/314 that caused your problem rather than this issue.

**Comment by wilkinsona** [comment]
I think this may be addressed by #330. I'll keep it open for now as a reminder to check the behaviour.
